File: backend/app/modules/task_management/tasks/schemas.py
import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Self

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.modules.task_management.domain.enums import Priority, TaskLinkType, TaskStatus, TaskType
from app.schemas.base import StrictRequestModel


class TaskCreateRequest(StrictRequestModel):
    parent_task_id: uuid.UUID | None = None
    task_type: TaskType = TaskType.TASK
    name: str = Field(min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=50_000)
    task_category: str | None = Field(default=None, max_length=100)
    assignee_id: uuid.UUID | None = None
    technical_lead_id: uuid.UUID | None = None
    functional_lead_id: uuid.UUID | None = None
    start_date: date | None = None
    end_date: date | None = None
    estimated_hours: Decimal = Field(default=Decimal("0"), ge=0, max_digits=10, decimal_places=2)
    priority: Priority = Priority.MEDIUM
    status: TaskStatus = TaskStatus.NEW
    blocked_by_id: uuid.UUID | None = None
    remarks: str | None = Field(default=None, max_length=20_000)

    @model_validator(mode="after")
    def validate_values(self) -> Self:
        self.name = self.name.strip()
        if not self.name:
            raise ValueError("name must not be blank")
        if self.start_date and self.end_date and self.end_date < self.start_date:
            raise ValueError("end_date must be on or after start_date")
        return self


class TaskUpdateRequest(StrictRequestModel):
    parent_task_id: uuid.UUID | None = None
    task_type: TaskType | None = None
    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=50_000)
    task_category: str | None = Field(default=None, max_length=100)
    assignee_id: uuid.UUID | None = None
    technical_lead_id: uuid.UUID | None = None
    functional_lead_id: uuid.UUID | None = None
    start_date: date | None = None
    end_date: date | None = None
    estimated_hours: Decimal | None = Field(default=None, ge=0, max_digits=10, decimal_places=2)
    priority: Priority | None = None
    blocked_by_id: uuid.UUID | None = None
    remarks: str | None = Field(default=None, max_length=20_000)
    version: int = Field(ge=1)

    @model_validator(mode="after")
    def validate_values(self) -> Self:
        if "name" in self.model_fields_set:
            if self.name is None:
                raise ValueError("name cannot be null")
            self.name = self.name.strip()
            if not self.name:
                raise ValueError("name must not be blank")
        if "priority" in self.model_fields_set and self.priority is None:
            raise ValueError("priority cannot be null")
        if "task_type" in self.model_fields_set and self.task_type is None:
            raise ValueError("task_type cannot be null")
        if "estimated_hours" in self.model_fields_set and self.estimated_hours is None:
            raise ValueError("estimated_hours cannot be null")
        if self.start_date and self.end_date and self.end_date < self.start_date:
            raise ValueError("end_date must be on or after start_date")
        return self


class TaskTransitionRequest(StrictRequestModel):
    to_status: TaskStatus
    version: int = Field(ge=1)
    reason: str | None = Field(default=None, max_length=1000)


class TaskArchiveRequest(StrictRequestModel):
    version: int = Field(ge=1)


class TaskResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    tenant_id: uuid.UUID
    task_id: uuid.UUID
    project_id: uuid.UUID
    task_number: int
    display_key: str
    task_type: TaskType
    parent_task_id: uuid.UUID | None
    name: str
    description: str | None
    task_category: str | None
    assignee_id: uuid.UUID | None
    technical_lead_id: uuid.UUID | None
    functional_lead_id: uuid.UUID | None
    reporter_id: uuid.UUID | None
    created_by_user_id: uuid.UUID | None
    start_date: date | None
    end_date: date | None
    estimated_hours: Decimal
    actual_hours: Decimal = Decimal("0")
    priority: Priority
    status: TaskStatus
    blocked_by_id: uuid.UUID | None
    remarks: str | None
    version: int
    created_at: datetime
    updated_at: datetime
    completed_at: datetime | None
    archived_at: datetime | None


class TaskLinkCreateRequest(StrictRequestModel):
    target_task_id: uuid.UUID
    link_type: TaskLinkType


class TaskLinkResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    link_id: uuid.UUID
    source_task_id: uuid.UUID
    target_task_id: uuid.UUID
    link_type: TaskLinkType
    created_by_user_id: uuid.UUID
    created_at: datetime

File: frontend/src/features/task-management/model/schemas.ts
import { z } from "zod";

import {
  LINK_TYPES,
  MEMBER_ROLES,
  PRIORITIES,
  PROJECT_STATUSES,
  TASK_STATUSES,
  TASK_TYPES,
} from "./types";

const id = z.string().uuid();
const nullableId = id.nullable();
const date = z.string().nullable();
const decimal = z.coerce.number().finite();

export const projectSchema = z.object({
  tenant_id: id,
  project_id: id,
  project_key: z.string(),
  name: z.string(),
  client_name: z.string().nullable(),
  description: z.string().nullable(),
  start_date: date,
  expected_end_date: date,
  status: z.enum(PROJECT_STATUSES),
  priority: z.enum(PRIORITIES),
  pm_id: nullableId,
  dm_id: nullableId,
  remarks: z.string().nullable(),
  version: z.number().int().positive(),
  created_at: z.string(),
  updated_at: z.string(),
  archived_at: z.string().nullable(),
});

export const projectMemberSchema = z.object({
  membership_id: id,
  tenant_id: id,
  project_id: id,
  user_id: id,
  role: z.enum(MEMBER_ROLES),
  added_by_user_id: nullableId,
  created_at: z.string(),
  updated_at: z.string(),
});

export const taskSchema = z.object({
  tenant_id: id,
  task_id: id,
  project_id: id,
  task_number: z.number().int().positive(),
  display_key: z.string(),
  task_type: z.enum(TASK_TYPES),
  parent_task_id: nullableId,
  name: z.string(),
  description: z.string().nullable(),
  task_category: z.string().nullable(),
  assignee_id: nullableId,
  technical_lead_id: nullableId,
  functional_lead_id: nullableId,
  reporter_id: nullableId,
  created_by_user_id: nullableId,
  start_date: date,
  end_date: date,
  estimated_hours: decimal,
  actual_hours: decimal,
  priority: z.enum(PRIORITIES),
  status: z.enum(TASK_STATUSES),
  blocked_by_id: nullableId,
  remarks: z.string().nullable(),
  version: z.number().int().positive(),
  created_at: z.string(),
  updated_at: z.string(),
  completed_at: z.string().nullable(),
  archived_at: z.string().nullable(),
});

export const taskLinkSchema = z.object({
  link_id: id,
  source_task_id: id,
  target_task_id: id,
  link_type: z.enum(LINK_TYPES),
  created_by_user_id: id,
  created_at: z.string(),
});

export const commentSchema = z.object({
  comment_id: id,
  task_id: id,
  commented_by_user_id: id,
  comment_text: z.string(),
  version: z.number().int().positive(),
  created_at: z.string(),
  updated_at: z.string(),
  deleted_at: z.string().nullable(),
});

export const timeEntrySchema = z.object({
  log_id: id,
  task_id: id,
  updated_by_user_id: id,
  hours_worked: decimal,
  work_date: z.string(),
  progress_notes: z.string().nullable(),
  version: z.number().int().positive(),
  log_date: z.string(),
  updated_at: z.string(),
  deleted_at: z.string().nullable(),
});

export const attachmentSchema = z.object({
  attachment_id: id,
  task_id: id,
  original_filename: z.string(),
  media_type: z.string(),
  size_bytes: z.number().int().nonnegative(),
  uploaded_by_user_id: id,
  created_at: z.string(),
});

export const activitySchema = z.object({
  event_id: id,
  task_id: id,
  event_type: z.string(),
  actor_user_id: nullableId,
  data: z.record(z.unknown()),
  occurred_at: z.string(),
});

export const userSummarySchema = z.object({
  tenant_id: id,
  user_id: id,
  name: z.string(),
  username: z.string().optional(),
  email: z.string(),
  role: z.string(),
  status: z.string(),
  version: z.number().int().positive(),
  created_by_user_id: nullableId,
  created_at: z.string(),
});

export const pageSchema = <T extends z.ZodTypeAny>(item: T) =>
  z.object({
    items: z.array(item),
    page: z.number().int().positive(),
    page_size: z.number().int().positive(),
    total: z.number().int().nonnegative(),
  });
